src/transformers/retrieval_rag.py

```diff
@@ -0,0 +1,337 @@
+        """
+
+        logger.info("initializing retrieval")
+
+        # initializing a separate process group for retrievel as the default
+        # nccl backend doesn't support gather/scatter operations while gloo
+        # is too slow to replace nccl for the core gpu communication
+        if dist.is_initialized():
+            logger.info("dist initialized")
+            # needs to be set manually
+            os.environ["GLOO_SOCKET_IFNAME"] = self._infer_socket_ifname()
+            # avoid clash with the NCCL port
+            os.environ["MASTER_PORT"] = str(distributed_port + 1)
+            self.process_group = dist.new_group(ranks=None, backend="gloo")
+
+        # initialize retriever only on the main worker
+        if not dist.is_initialized() or self._is_main():
+            logger.info("dist not initialized / main")
+            self.retriever.init_index()
+
+        # all processes wait untill the retriever is initialized by the main process
+        if dist.is_initialized():
+            torch.distributed.barrier(group=self.process_group)
+
+    def _is_main(self):
+        return dist.get_rank(group=self.process_group) == 0
+
+    def _chunk_tensor(self, t, chunk_size):
+        n_chunks = t.shape[0] // chunk_size + int(t.shape[0] % chunk_size > 0)
+        return list(torch.chunk(t, n_chunks, dim=0))
+
+    def _scattered(self, scatter_list, target_shape, target_type=torch.float32):
+        target_tensor = torch.empty(target_shape, dtype=target_type)
+        dist.scatter(target_tensor, src=0, scatter_list=scatter_list, group=self.process_group)
+        return target_tensor
+
+    def _infer_socket_ifname(self):
+        addrs = psutil.net_if_addrs()
+        # a hacky way to deal with varying network interface names
+        ifname = next((addr for addr in addrs if addr.startswith("e")), None)
+        return ifname
+
+    def _main_retrieve(self, query_vectors):
+        query_vectors_batched = self._chunk_tensor(query_vectors, self.batch_size)
+        ids_batched = []
+        vectors_batched = []
+        for query_vectors in query_vectors_batched:
+            start_time = time.time()
+            ids, vectors = self.retriever.get_top_docs(query_vectors.numpy(), self.n_docs)
+            logger.debug(
+                "index search time: {} sec, batch size {}".format(time.time() - start_time, query_vectors.shape)
+            )
+            ids_batched.append(ids)
+            vectors_batched.append(vectors)
+        return torch.cat(ids_batched), torch.cat(vectors_batched)
+
+    def _finalize_retrieval(self, query_vectors, doc_ids, doc_vectors):
+        assert doc_ids.shape[1] == self.n_docs and doc_vectors.shape[1] == self.n_docs
+        doc_vectors = doc_vectors.to(query_vectors)
+        doc_scores = torch.bmm(query_vectors.unsqueeze(1), doc_vectors.transpose(1, 2)).squeeze(1)
+        doc_dicts = self.retriever.get_doc_dicts(doc_ids)
+        return doc_scores, doc_dicts
```

Comment:
Indeed, this part is framework specific.

But couldn't `query_vectors ` and `doc_vectors` be in the model actually?

I don't see any interaction between {`query_vectors ` and `doc_vectors`} on one hand and {`doc_ids` and `doc_dicts`} on the other hand and the later seems to be the real elements related to the Retriever.

Reply:
I mean, in general I don't think we expect to back-propagate in the retriever component it-self (at least in the short term research horizon 🙂) so we might be able to decouple both elements.

Reply:
> But couldn't query_vectors and doc_vectors be in the model actually?

Yup we could do that - that would be the second option I described.

Reply:
Oh yes indeed. Ok I think this would be a nice option, this way we can think about making a TensorFlow or JAX version of the model and we'll be able to share the retriever between both (like we do for the tokenizers).